crates/sound/src/audio_backends/pw_backend.rs

```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
 
     fn set_param(&self, stream_id: u32, params: PCMParams) -> Result<()> {
         let mut stream_params = self.stream_params.write().unwrap();
+        let mut stream_states = self.stream_states.write().unwrap();
         stream_params[stream_id as usize] = params;
+        // state {SET_PARAM} -> {SET_PARAM}
+        // state {PREPARE} -> {PREPARE}
+        if stream_states[stream_id as usize] == 0 {
+            // state {0} -> {SET_PARAMS}
+            stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_SET_PARAMS
+        }
         Ok(())
     }
 
-    fn prepare(&self, _stream_id: u32) -> Result<()> {
+    fn prepare(&self, stream_id: u32) -> Result<()> {
+        let mut stream_states = self.stream_states.write().unwrap();
+        let mut stream_hash = self.stream_hash.write().unwrap();
         self.thread_loop.lock();
-        self.thread_loop.unlock();
+        let stream_params = self.stream_params.read().unwrap();
+        let params = stream_params[stream_id as usize].clone();
+        //dbg!("params: {}", params);
+
+        if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_SET_PARAMS {
+            let mut buff = [0; 1024];
+            let p_buff = &mut buff as *mut i32 as *mut c_void;
+
+            let mut b = spa_pod_builder {
+                data: p_buff,
+                size: buff.len() as u32,
+                _padding: 0,
+                callbacks: spa_callbacks {
+                    funcs: std::ptr::null(),
+                    data: std::ptr::null_mut(),
+                },
+                state: spa_pod_builder_state {
+                    offset: 0,
+                    flags: 0,
+                    frame: std::ptr::null_mut(),
+                },
+            };
+
+            let mut pos: [u32; 64] = [0; 64];
+
+            //todo map position to chmap
+            pos[0] = SPA_AUDIO_CHANNEL_FL;
+            pos[1] = SPA_AUDIO_CHANNEL_FR;
+            let mut info = spa_audio_info_raw {
+                format: match params.format {
+                    VIRTIO_SND_PCM_FMT_S8 => SPA_AUDIO_FORMAT_S8,
```

Comment:
I feel like those `VIRTIO_SND_PCM_FMT_` constants should be an enum instead and then implement some function to give you the `SPA_AUDIO_FORMAT_` equivalent. Ideally the later would also be a proper enum exposed by the `libspa` bindings from upstream.

Reply:
yes, we could use enum. /me thinks about it.

Reply:
I agree it's best to put them in the libspa bindings upstream. Thanks :)

Reply:
But  `VIRTIO_SND_PCM_FMT_` comes from virtio-sound

Reply:
yes it comes from virtio-sound, and other parts of the virtio-sound uses those constants, I will prefer to leave it as constants.